onnxruntime/python/onnxruntime_pybind_state.cc

```diff
@@ -613,13 +613,32 @@ void addObjectMethods(py::module& m, Environment& env) {
                      DataTypeImpl::GetType<Tensor>()->GetDeleteFunc());
         auto status = io_binding->Get()->BindOutput(name, mlvalue);
         if (!status.IsOK())
-          throw std::runtime_error("Error when bind input: " + status.ErrorMessage());
+          throw std::runtime_error("Error when bind output: " + status.ErrorMessage());
+      })
+      .def("bind_output_name", [](SessionIOBinding* io_binding, const std::string& name) -> void {
+        OrtValue mlvalue;
+        auto status = io_binding->Get()->BindOutput(name, mlvalue);
```

Comment:
Not quite understanding how this works. 

I expected you would have to add a feature to ORT to be able to say what device you want an output on, so that any post-execution copy logic can take that into account. Otherwise it would run the graph and copy to the default location of OrtValue which is CPU I believe.

What aspect of this mechanism provides the info that you want the output to be on a specific device?

Reply:
this api is a bit non-intuitive. what does it mean to bind just an output name? 
i think what you're trying to accomplish is a way to only bind input to gpu device but not the output? could it have been implemented in a cleaner way where you don't need to introduce new specific api's? 
and should it be generically extended to support the other case where you bind output but not input? 

+@souptc who implemented this api for training to see if this api change makes sense.

Reply:
> Not quite understanding how this works.
> 
> I expected you would have to add a feature to ORT to be able to say what device you want an output on, so that any post-execution copy logic can take that into account. Otherwise it would run the graph and copy to the default location of OrtValue which is CPU I believe.
> 
> What aspect of this mechanism provides the info that you want the output to be on a specific device?

ORT already supports specifying where the output should be. By default, the output will be put on CPU. This mechanism just allows users to only bind input to gpu device but not the output as Georgy mentioned in his comments.

Reply:
> this api is a bit non-intuitive. what does it mean to bind just an output name?
> i think what you're trying to accomplish is a way to only bind input to gpu device but not the output? could it have been implemented in a cleaner way where you don't need to introduce new specific api's?
> and should it be generically extended to support the other case where you bind output but not input?
> 
> +@souptc who implemented this api for training to see if this api change makes sense.

yes, I'm trying to do exactly what you said: bind input to gpu but not output. If this api is not intuitive, an alternative is to add a boolean parameter in bind_input Api: bind_input_only so that we can get rid of bind_output_name API. Would that make it better?

Regard binding output but not input, I'm not sure we have this use case. We need to bind input only because we only know the input shape in some cases. Is there any case in which we only know output shape but not the input shape? If so, we can add the feature of binding output only.

Reply:
i think the issue here is the api becomes inconsistent. 
the existing python iobinding api says you need to bind to a memory buffer on device X.
but here you introduce a method that binds output but only gets passed a name. 
it's difficult for caller to rationalize what that means relative to the existing api's.  
i think this is because there some difference between the underlying C++ api and what is exposed to python?
@pranavsharma , @souptc any thoughts on this?

Reply:
Could we add an optional output_names parameter to run_with_iobinding instead? That may be more intuitive (if you don't explicitly bind outputs, provide output_names when running).

Reply:
> Could we add an optional output_names parameter to run_with_iobinding instead? That may be more intuitive (if you don't explicitly bind outputs, provide output_names when running).

Would that imply that run_with_iobinding() then would fetch the outputs associated with those names if output_names are provided? 
or would one still use iobinding.get_outputs() ?

or what one made the parameters after name (e.g. device) optional for iobinding.bind_output() instead of introducing a new bind_output_names() method.

Reply:
> > Could we add an optional output_names parameter to run_with_iobinding instead? That may be more intuitive (if you don't explicitly bind outputs, provide output_names when running).
> 
> Would that imply that run_with_iobinding() then would fetch the outputs associated with those names if output_names are provided?
> or would one still use iobinding.get_outputs() ?
we probably still need to use get_outputs().

> 
> or what if one made the parameters after name (e.g. device) optional for iobinding.bind_output() instead of introducing a new bind_output_names() method.

I thought about it which is certainly feasible. However, it's a bit hard to express the contract  that the parameters of bind_output() except output_names should be either all or none.

Reply:
I think bind_output having optional parameters for device, etc would be much better. This should be the same for bind_input as well for symmetry. This would indicate that you're choosing the defaults (CPU). And outputs can be obtained using get_outputs.

Reply:
ORT doesn't (currently) have the notion of allocating the output returned to the user on a specific device if a pre-allocated output is not provided. It's always on CPU. If it add that it should be possible to just specify name and device when binding an output. It would also support a use case where you want to keep the output on a different device but don't know the output shape upfront. 

Pending that the bind_output api could change to allow just the name and device to be provided, but fail for anything other than CPU as the device.

Reply:
> I think bind_output having optional parameters for device, etc would be much better. This should be the same for bind_input as well for symmetry. This would indicate that you're choosing the defaults (CPU). And outputs can be obtained using get_outputs.

Agree. Let's go this way.

Reply:
> ORT doesn't (currently) have the notion of allocating the output returned to the user on a specific device if a pre-allocated output is not provided. It's always on CPU. If it add that it should be possible to just specify name and device when binding an output. It would also support a use case where you want to keep the output on a different device but don't know the output shape upfront.
> 
> Pending that the bind_output api could change to allow just the name and device to be provided, but fail for anything other than CPU as the device.


It's an interesting idea. @pranavsharma and @jywu-msft, if you guys agree to do it, I can do that in another PR.

Reply:
> > ORT doesn't (currently) have the notion of allocating the output returned to the user on a specific device if a pre-allocated output is not provided. It's always on CPU. If it add that it should be possible to just specify name and device when binding an output. It would also support a use case where you want to keep the output on a different device but don't know the output shape upfront.
> > Pending that the bind_output api could change to allow just the name and device to be provided, but fail for anything other than CPU as the device.
> 
> It's an interesting idea. @pranavsharma and @jywu-msft, if you guys agree to do it, I can do that in another PR.

Yes, i think you can look at enabling the use case Scott mentioned in another PR.

Reply:
@jywu-msft @pranavsharma and @skottmckay  I updated the API, please review it. Thanks.